hbase-server/src/test/java/org/apache/hadoop/hbase/master/TestMasterShutdown.java

```diff
@@ -151,19 +156,42 @@ public void testMasterShutdownBeforeStartingAnyRegionServer() throws Exception {
       hbaseCluster = new LocalHBaseCluster(htu.getConfiguration(), options.getNumMasters(),
         options.getNumRegionServers(), options.getMasterClass(), options.getRsClass());
       final MasterThread masterThread = hbaseCluster.getMasters().get(0);
+
       masterThread.start();
-      // Switching to master registry exacerbated a race in the master bootstrap that can result
-      // in a lost shutdown command (HBASE-8422, HBASE-23836). The race is essentially because
-      // the server manager in HMaster is not initialized by the time shutdown() RPC (below) is
-      // made to the master. The suspected reason as to why it was uncommon before HBASE-18095
-      // is because the connection creation with ZK registry is so slow that by then the server
-      // manager is usually init'ed in time for the RPC to be made. For now, adding an explicit
-      // wait() in the test, waiting for the server manager to become available.
-      final long timeout = TimeUnit.MINUTES.toMillis(10);
-      assertNotEquals("Timeout waiting for server manager to become available.",
-        -1, Waiter.waitFor(htu.getConfiguration(), timeout,
-          () -> masterThread.getMaster().getServerManager() != null));
-      htu.getConnection().getAdmin().shutdown();
+      final CompletableFuture<Void> shutdownFuture = CompletableFuture.runAsync(() -> {
+        // Switching to master registry exacerbated a race in the master bootstrap that can result
+        // in a lost shutdown command (HBASE-8422, HBASE-23836). The race is essentially because
+        // the server manager in HMaster is not initialized by the time shutdown() RPC (below) is
+        // made to the master. The suspected reason as to why it was uncommon before HBASE-18095
+        // is because the connection creation with ZK registry is so slow that by then the server
+        // manager is usually init'ed in time for the RPC to be made. For now, adding an explicit
+        // wait() in the test, waiting for the server manager to become available.
+        final long timeout = TimeUnit.MINUTES.toMillis(10);
+        assertNotEquals("timeout waiting for server manager to become available.", -1,
+          htu.waitFor(timeout, () -> masterThread.getMaster().getServerManager() != null));
+
+        // Master has come up far enough that we can terminate it without creating a zombie.
+        final long result = htu.waitFor(timeout, 1000, () -> {
+          LOG.debug("Attempting to establish connection.");
+          try (final Connection conn = htu.getConnection()) {
+            conn.getAdmin().shutdown();
+            LOG.info("Shutdown RPC sent.");
+            return true;
+          } catch (IOException|CompletionException e) {
+            LOG.error("Failed to establish connection.");
+            if (connectionFailedWithMaster(e)) {
```

Comment:
I think this is a bit hacky, instead, can we just issue one shutdown(), don't worry about the result of it (since we know it is going to succeed even if it runs into ZK issues) and then just join on the master thread? That means the code will look like this..

master.start()
htu.waitFor(.... masterThread.getMaster().serverMgr() != null);
Thread.run() {
  conn.getAdmin().shutdown();
 // don't worry about the return value
}
masterThread.join()
shutdownThread.join()

Thoughts?

Reply:
This is what I thought of doing first thing, yesterday :)
Then I felt better to check the exact error message and confirm. Although we are sure at this point about the failure, I don't have any strong opinion against not validating the exact error message. Let's do this as you mentioned.